### ISSUE TITLE
using File printer and catching IOError

### ISSUE THREAD

**Issue by eccentric** [python-escpos/python-escpos#189]
I ran into some issues accessing a file device which was raising an IOError due to a permissions problem using version 2.2.0. Nothing out of the ordinary, and I thought I'd write a try statement to catch it but ended up with another exception after handling the IOError. Example:

```
from escpos import printer
try:
    p = printer.File('/dev/missing')
except IOError, e:
    print('ioerror', e)
```
```
('ioerror', IOError(13, 'Permission denied'))
Exception AttributeError: "'NoneType' object has no attribute 'flush'" in <bound method File.`__del__` of <escpos.printer.File object at 0x7f36b9647c50>> ignored
```
This seems to be caused by the garbage collection, calling the `__del__` method which calls self.close() which then runs self.device.flush(), but device is None. The close function on the printer.File class should maybe check that the value of device is not None before calling flush/close? (seems to be what is done with usb/serial printer classes).

I checked out the latest development code, and realized this error doesn't happen, but the exact same `__del__` method and close function are present, and it seems like `__del__` just does't get called. Looking into it some more, I think it was changed as of commit a3ca2c2a16fbba3592dd9d00818efb895939d208 which added a circular reference with MagicEncode keeping reference to self.device (which is the Escpos class). And so I think the garbage collection is not ever calling the `__del__` method since that change. I don't really understand the garbage collection to know if that is a problem or not, but thought I'd mention it in case it is a problem.

**Comment by patkan** [comment]
Hi, thank you for the close examination!
So if I understand right two things have to be checked and fixed:
* [ ] is the garbage collection broken by MagicEncode
* [ ] the existence of the File object should be checked before trying to delete it

**Comment by eccentric** [comment]
Yes, I'm not sure if the garbage collection is an issue or not, but it does seem that the `__del__` function is not being called after the class is no longer used or removed since that commit. Which may or may not be the correct behaviour.

The printer File object, just needs the close method to be updated to check `if self.device is not None:` before calling `self.device.flush()` and `self.device.close()`, which fixes the issue with the second exception I was having after capturing the IOError. (just like the Serial or Usb printer classes have).

**Comment by belono** [comment]
@patkan this issue was fixed at #201.
The issue is no longer reproducible since v.3.0a so you can close as completed.
